src/map/treasure_pool_container.cpp

```diff
@@ -0,0 +1,225 @@
+    }
+
+    // 3. Part of an alliance
+    if (PChar->PParty &&
+        PChar->PParty->m_PAlliance &&
+        PChar->PParty->m_PAlliance->getMainParty() &&
+        PChar->PParty->m_PAlliance->getMainParty()->GetLeader())
+    {
+        const auto PAllianceLeader = PChar->PParty->m_PAlliance->getMainParty()->GetLeader();
+
+        // 3a. Find any pool "owned" by the alliance leader, even if they're not in the zone
+        if (m_Pools.contains(PAllianceLeader->id))
+        {
+            auto& PTreasurePool = m_Pools.at(PAllianceLeader->id);
+            if (!PTreasurePool.IsMember(PChar))
+            {
+                PTreasurePool.AddMember(PChar);
+            }
+            return PTreasurePool;
+        }
+
+        // 3b. First alliance member in zone
+        auto& newTp = CreateTreasurePool(PAllianceLeader->id, TREASUREPOOL_ALLIANCE, TREASUREPOOL_MANAGED);
+        newTp.AddMember(PChar);
+        return newTp;
+    }
+
+    // 4. Part of a party
+    if (PChar->PParty && PChar->PParty->GetLeader())
+    {
+        const auto PPartyLeader = PChar->PParty->GetLeader();
+
+        // 4a. Find any pool "owned" by the party leader, even if they're not in the zone
+        if (m_Pools.contains(PPartyLeader->id))
+        {
+            auto& PTreasurePool = m_Pools.at(PPartyLeader->id);
+            if (!PTreasurePool.IsMember(PChar))
+            {
+                PTreasurePool.AddMember(PChar);
+            }
+            return PTreasurePool;
+        }
+
+        // 4b. First party member in zone
+        auto& newTp = CreateTreasurePool(PPartyLeader->id, TREASUREPOOL_PARTY, TREASUREPOOL_MANAGED);
+        newTp.AddMember(PChar);
+        return newTp;
+    }
+
+    // 5. Find any solo pool we may have already created
+    if (m_Pools.contains(PChar->id))
+    {
+        return m_Pools.at(PChar->id);
+    }
+
+    // Otherwise, create a new solo pool
+    auto& newTp = CreateTreasurePool(PChar->id, TREASUREPOOL_SOLO, TREASUREPOOL_MANAGED);
+    newTp.AddMember(PChar);
+    return newTp;
+}
+
+auto CTreasurePoolContainer::GetTreasurePools() -> std::unordered_map<uint32, CTreasurePool>&
+{
+    return m_Pools;
+}
+
+// Reassigns the map key for a given treasure pool.
+// Used when party or alliance leader changes.
+auto CTreasurePoolContainer::ReassignTreasurePool(const CCharEntity* PPrev, const CCharEntity* PNew) -> bool
+{
+    if (!PPrev || !PNew)
+    {
+        ShowError("CTreasurePoolContainer::ReassignTreasurePool() - PPrev or PNew was null.");
+        return false;
+    }
+
+    if (auto node = m_Pools.extract(PPrev->id); !node.empty())
+    {
+        node.key() = PNew->id;
+        m_Pools.insert(std::move(node));
+        return true;
+    }
+
+    return false;
+}
+
+// Force releases a treasure pool from the container.
+// Used when dissolving a party or alliance.
+auto CTreasurePoolContainer::ReleaseTreasurePool(CTreasurePool& PTreasurePool) -> void
+{
+    for (auto it = m_Pools.begin(); it != m_Pools.end();)
```

Comment:
Same here - as a general rule this isn't the hot-path so we don't need to worry about the mega/micro performance benefits of this type of iteration